### PR TITLE
[cdc_rsync] Improve throughput for local copies

### PR DESCRIPTION
**Pull request by ljusten**
On Windows, fclose() seems to be very expensive for large files, where closing a 1 GB file takes up to 5 seconds. This CL calls fclose() in background threads. This tremendously improves local syncs, e.g. copying a 4.5 GB, 300 files data set takes only 7 seconds instead of 30 seconds.

Also increases the buffer size for copying from 16K to 128K (better throughput for local copies), and adds a timestamp to debug and verbose console logs (useful when comparing client and server logs).

### CODE REVIEW

File: cdc_rsync_server/cdc_rsync_server.cc [hunk 1]
```diff
@@ -126,11 +164,78 @@ class PatchTask : public Task {
...
+  // Finalize |file| with given path |filepath|. |status| is the status from
+  // writing the file. On error, the file is only closed.
+  FinalizeCopiedFileTask(FILE* fp, FileInfo file, std::string filepath,
+                         bool is_executable, absl::Status status)
+      : fp_(fp),
+        file_(std::move(file)),
+        filepath_(std::move(filepath)),
+        is_executable_(is_executable),
+        status_(status) {}
+  virtual ~FinalizeCopiedFileTask() = default;
```
Comment (beckerhe) [inline cdc_rsync_server/cdc_rsync_server.cc:197]: Why is this line needed? I would expect `Task` to have a virtual destructor and if not it should probably have one. And then there is no need to redeclare if virtual here.

File: cdc_rsync_server/cdc_rsync_server.cc [hunk 2]
```diff
@@ -126,11 +164,78 @@ class PatchTask : public Task {
     status_ = path::SetFileTime(target_filepath_, file_.client_modified_time);
   }
 
+  const std::string base_filepath_;
+  const std::string target_filepath_;
+  const ChangedFileInfo file_;
+  CdcInterface* const cdc_;
+  const bool need_intermediate_file_ = false;
+  const std::string patched_filepath_;
```
Comment (beckerhe) [inline cdc_rsync_server/cdc_rsync_server.cc:172]: Declaring those fields `const` will disable the move constructor and move assignment operator. Not sure if that's relevant to you though.

File: cdc_rsync_server/cdc_rsync_server.cc [hunk 3]
```diff
@@ -126,11 +164,78 @@ class PatchTask : public Task {
...
+
  private:
-  std::string base_filepath_;
-  std::string target_filepath_;
-  ChangedFileInfo file_;
-  CdcInterface* cdc_;
+  FILE* const fp_ = nullptr;
+  const FileInfo file_;
+  const std::string filepath_;
+  const bool is_executable_;
```
Comment (beckerhe) [inline cdc_rsync_server/cdc_rsync_server.cc:238]: Same as above.

File: cdc_rsync_server/cdc_rsync_server.cc [hunk 4]
```diff
@@ -152,7 +257,7 @@ PathFilter::Rule::Type ToInternalType(
 
 }  // namespace
 
-CdcRsyncServer::CdcRsyncServer() = default;
+CdcRsyncServer::CdcRsyncServer() {}
```
Comment (beckerhe) [inline cdc_rsync_server/cdc_rsync_server.cc:260]: Why this change?

The `= default;` style has no disadvantages and should be the default because it keeps the default constructor trivial if possible. (Not really relevant here though, because the class has members with non-trivial default constructors, so this constructor will also not be trivial. So feel free to ignore.)

File: cdc_rsync_server/cdc_rsync_server.cc [hunk 5]
```diff
@@ -619,27 +726,25 @@ absl::Status CdcRsyncServer::HandleSendMissingFileData() {
...
+    finalize_pool.QueueTask(std::make_unique<FinalizeCopiedFileTask>(
+        *fp, file, filepath, is_executable, status));
+    finalize_pool.WaitForQueuedTasksAtMost(kMaxQueueSize);
 
-    // Set file write time.
-    status = path::SetFileTime(filepath, file.modified_time);
-    if (!status.ok()) {
-      return WrapStatus(status, "Failed to set file mod time for %s", filepath);
+    // Drain finalize pool for the last file.
+    if (server_index + 1 == diff_.missing_files.size()) {
```
Comment (beckerhe) [inline cdc_rsync_server/cdc_rsync_server.cc:734]: Nit: The name `server_index` is confusing to me. It took me a while to figure out that it is the index into `missing_files`. Since this is technically not related to this change, feel free to do with that information what you want. :-)

File: cdc_rsync_server/cdc_rsync_server.cc [hunk 6]
```diff
@@ -619,27 +726,25 @@ absl::Status CdcRsyncServer::HandleSendMissingFileData() {
...
-      status = path::GetStats(filepath, &stats);
-      if (status.ok()) {
-        status = path::ChangeMode(filepath, stats.mode | kExecutableBits);
-      }
-      if (!status.ok()) {
-        LOG_WARNING("Failed to set executable bit on '%s': %s", filepath,
-                    status.ToString());
+    // Check the results of completed tasks.
+    std::unique_ptr<Task> task;
+    while ((task = finalize_pool.TryGetCompletedTask()) != nullptr) {
```
Comment (beckerhe) [inline cdc_rsync_server/cdc_rsync_server.cc:740]: Nit: Why not a for loop? Wouldn't a for loop be better suited here?

File: cdc_rsync_server/cdc_rsync_server.cc [hunk 7]
```diff
@@ -619,27 +726,25 @@ absl::Status CdcRsyncServer::HandleSendMissingFileData() {
...
+      const FinalizeCopiedFileTask* finalize_task =
+          static_cast<FinalizeCopiedFileTask*>(task.get());
+      if (!finalize_task->Status().ok()) {
+        // Close and finish files that have already been copied, so we don't
+        // several already copied files because one failed.
```
Comment (beckerhe) [inline cdc_rsync_server/cdc_rsync_server.cc:745]: There seems to be a verb missing in the comment.

File: cdc_rsync_server/cdc_rsync_server.cc [hunk 8]
```diff
@@ -708,25 +824,28 @@ absl::Status CdcRsyncServer::SyncChangedFiles() {
...
+      patch_pool.Wait();
+      finalize_pool.Wait();
     }
 
     // Check the results of completed tasks.
-    std::unique_ptr<Task> task = pool.TryGetCompletedTask();
-    while (task) {
-      PatchTask* patch_task = static_cast<PatchTask*>(task.get());
+    std::unique_ptr<Task> task;
+    while ((task = finalize_pool.TryGetCompletedTask()) != nullptr) {
```
Comment (beckerhe) [inline cdc_rsync_server/cdc_rsync_server.cc:838]: Nit: Same as above (for loop)

File: common/threadpool.cc [hunk 9]
```diff
@@ -45,6 +45,11 @@ void Threadpool::Shutdown() {
   for (auto& worker : workers_) {
     if (worker.joinable()) worker.join();
   }
+
+  // Discard all completed tasks.
+  absl::MutexLock lock(&completed_tasks_mutex_);
+  std::queue<std::unique_ptr<Task>> empty;
+  std::swap(completed_tasks_, empty);
```
Comment (beckerhe) [inline common/threadpool.cc:52]: Why the swap? To me this is implying that more work needs to be done, while move assignment from a temporary default constructed queue should be enough:

 `completed_tasks_ = {};`

Or is this missing some return statement?

**Review by beckerhe — DISMISSED** [review]
LGTM. Nothing serious. Only some style related minor comments.

File: common/threadpool_test.cc [hunk 10]
```diff
@@ -151,5 +151,33 @@ TEST_F(ThreadpoolTest, GetCompletedTask) {
...
+}
+
+TEST_F(ThreadpoolTest, WaitForQueuedTasksAtMost) {
+  auto task_func = [](Task::IsCancelledPredicate) { Util::Sleep(10); };
+
+  Semaphore task_finished(0);
+  Threadpool pool(1);
+  pool.QueueTask(std::make_unique<TestTask>(task_func));
+  pool.QueueTask(std::make_unique<TestTask>(task_func));
+  pool.WaitForQueuedTasksAtMost(1);
```
[the remark this answers is not in the document]
Comment (ljusten) [reply common/threadpool_test.cc:178]: You're right, if more than 10 ms between `pool.WaitForQueuedTasksAtMost(1)` and the `EXPECT_EQ(pool.NumQueuedTasks(), 1);` pass, this would fail. Updated it.

File: cdc_rsync_server/cdc_rsync_server.cc (hunk 1 above)
Comment (ljusten) [reply cdc_rsync_server/cdc_rsync_server.cc:197]: Not strictly necessary, but I would consider it good practice. My main reason is, if you don't declare a destructor, it'll be inlined in the .h file (in this case the class is defined and declared in .cc, so irrelevant here). If you forward declare a type, and you have a std::unique_ptr of that type, this will fail since it won't know in the .h file how to destroy it. If you define the constructor as default in the .cc file, it will know how to destroy the pointer.

File: cdc_rsync_server/cdc_rsync_server.cc (hunk 2 above)
Comment (ljusten) [reply cdc_rsync_server/cdc_rsync_server.cc:172]: Tasks are unique ptrs that shouldn't be copied. Deleting copy constructors and assignment operators just to be sure.

File: cdc_rsync_server/cdc_rsync_server.cc (hunk 3 above)
Comment (ljusten) [reply cdc_rsync_server/cdc_rsync_server.cc:238]: Same as above.

File: cdc_rsync_server/cdc_rsync_server.cc (hunk 4 above)
Comment (ljusten) [reply cdc_rsync_server/cdc_rsync_server.cc:260]: I have no memories of this change. I probably added a constructor arg at some point and removed it. Reverted.

File: cdc_rsync_server/cdc_rsync_server.cc (hunk 5 above)
Comment (ljusten) [reply cdc_rsync_server/cdc_rsync_server.cc:734]: The terms `client_index` and `server_index` are used throughout the sync code. The former is an index into arrays of files on the client, whereas the latter is the equivalent for the server. Some code in the client deals with both at the same time, and this naming scheme keeps the code sane.

File: cdc_rsync_server/cdc_rsync_server.cc (hunk 7 above)
Comment (ljusten) [reply cdc_rsync_server/cdc_rsync_server.cc:745]: Done.

File: cdc_rsync_server/cdc_rsync_server.cc (hunk 6 above)
Comment (ljusten) [reply cdc_rsync_server/cdc_rsync_server.cc:740]: I guess now it's OK. In some iteration, TryGetCompletedTask() was inlined, so since the for loop has to call TryGetCompletedTask() twice, it was pretty ugly. Visually, I still prefer the while loop, but it gets negative points for inlined assignment + check.

File: cdc_rsync_server/cdc_rsync_server.cc (hunk 8 above)
Comment (ljusten) [reply cdc_rsync_server/cdc_rsync_server.cc:838]: Done.

File: common/threadpool.cc (hunk 9 above)
Comment (ljusten) [reply common/threadpool.cc:52]: Swapping deletes memory, whereas the move assignment doesn't. Not really super relevant, since Shutdown() is right now always called when the pool is deleted, anyway.

```
  std::vector<int> vec;
  vec.resize(100, 123);
  printf("%zu\n", vec.capacity());
  vec = {};
  printf("%zu\n", vec.capacity());
  std::vector<int> empty;
  std::swap(vec, empty);
  printf("%zu\n", vec.capacity());
```
outputs
```
100
100
0
```

**Review by beckerhe — APPROVED** [review]
